compiler/optimizer/optimizer.go

```diff
@@ -31,6 +32,46 @@ func (o *Optimizer) Entry() *dag.Sequential {
 	return o.entry
 }
 
+// MergeFilters transforms the DAG by merging adjacent filter operators so that,
+// e.g., "where a | where b" becomes "where a and b".
+//
+// Note: MergeFilters does not descend into dag.OverExpr.Scope, so it cannot
+// merge filters in "over" expressions like "sum(over a | where b | where c)".
+func (o *Optimizer) MergeFilters() { mergeFilters(o.entry) }
+
+func mergeFilters(op dag.Op) {
+	switch op := op.(type) {
+	case *dag.From:
+		for _, t := range op.Trunks {
+			mergeFilters(t.Seq)
+		}
+	case *dag.Over:
+		mergeFilters(op.Scope)
+	case *dag.Parallel:
+		for _, o := range op.Ops {
+			mergeFilters(o)
+		}
+	case *dag.Sequential:
```

Comment:
I am wondering if we want to add something like dag.Walk and use this here. We have now a few areas of optimizer that does this and as we add additional operators I'm concerned we'll miss the different places in optimizer that crawls over dags looking for leafs. For instance my user defined operators pr will need to do these kind of updates...

Reply:
This was a really good suggestion!

I didn't try to add a fully-fledged dag.Walk because I wasn't confident I'd get the interface right, but I think reimplementing what was here in terms of a bare-bones walkOp function made it a lot clearer, and we can expand walkOp into dag.Walk down the road.